scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/synchronization/AOSynchronization.java

```diff
@@ -814,28 +850,40 @@ public void runActivity(Body body) {
-                Request request = requestWithWaitTime.getNewRequest();
-                if (request != null && request.getMethodName().equals("freeze")) {
-                    service.serve(request);
-                    service.blockingServeOldest("resume");
-                } else if (request != null && request.getMethodName().startsWith("wait") &&
-                           !testWaitFunction(service, request)) {
-                    // If the predicate is not met, delay the wait method execution
-                    TimedOutRequest timedOutRequest = new TimedOutRequest(request,
-                                                                          extractWaitRequestTimeoutParameter(request));
-                    logger.trace("New pending wait request : " + timedOutRequest);
-                    waitUntilQueue.add(new TimedOutRequest(request, extractWaitRequestTimeoutParameter(request)));
-                } else if (request != null) {
-                    service.serve(request);
+                List<NewRequestWithWaitTime> requestsWithWaitTime = waitForNewRequests(service);
+                for (NewRequestWithWaitTime requestWithWaitTime : requestsWithWaitTime) {
+                    Request request = requestWithWaitTime.getNewRequest();
+                    if (request != null && request.getMethodName().equals("freeze")) {
+                        service.serve(request);
+                        service.blockingServeOldest("resume");
+                    } else if (request != null && request.getMethodName().startsWith("wait") &&
+                               !testWaitFunction(service, request)) {
+                        // If the predicate is not met, delay the wait method execution
+                        TimedOutRequest timedOutRequest = new TimedOutRequest(request,
+                                                                              extractWaitRequestTimeoutParameter(request));
+                        logger.trace("New pending wait request : " + timedOutRequest);
+                        waitUntilQueue.add(new TimedOutRequest(request, extractWaitRequestTimeoutParameter(request)));
+                    } else if (request != null) {
+                        service.serve(request);
+                    }
                 }
-                unblockWaitMethods(service, requestWithWaitTime.getWaitTime());
+                unblockWaitMethods(service, computeMaxTimeSpentWaiting(requestsWithWaitTime));
             } catch (InterruptedException e) {
                 Thread.currentThread().interrupt();
             }
         }
     }
 
+    private long computeMaxTimeSpentWaiting(List<NewRequestWithWaitTime> requestsWithWaitTime) {
+        long maxWait = 0;
+        for (NewRequestWithWaitTime requestWithWaitTime : requestsWithWaitTime) {
+            if (requestWithWaitTime.waitTime > maxWait) {
+                maxWait = requestWithWaitTime.waitTime;
+            }
```

Comment:
Can be enhanced like this:
maxWait = requestWithWaitTime.waitTime > maxWait ? requestWithWaitTime.waitTime : maxWait;

Reply:
sorry nebil, I did not apply this suggestion as I want to merge this asap